certbot/plugins/dns_cloudflare.py

```diff
@@ -0,0 +1,323 @@
+
+        if self.email:
+            setattr(self.config, self.dest('email'), self.email)
+        else:
+            raise errors.PluginError('Cloudflare account email address required to proceed.')
+
+        configured_api_key = self.conf('api-key')
+        if configured_api_key:
+            self.api_key = configured_api_key
+        else:
+            self.api_key = self._prompt_for_data('API key')
+
+        if self.api_key:
+            setattr(self.config, self.dest('api-key'), self.api_key)
+        else:
+            raise errors.PluginError('Cloudflare account API key required to proceed.')
+
+    @staticmethod
+    def _prompt_for_data(label):
+        """
+        Prompt the user for a piece of information.
+
+        :param string label: The user-friendly label for this piece of information.
+        :returns: The user's response (guaranteed non-empty).
+        :rtype: string
+        """
+
+        display = zope.component.getUtility(interfaces.IDisplay)
+
+        while True:
+            code, response = display.input(
+                'Input Cloudflare account {0}'.format(label),
+                force_interactive=True)
+            if code == display_util.HELP:
+                # Displaying help is not currently implemented
+                return None
+            elif code == display_util.CANCEL:
+                return None
+            else:  # code == display_util.OK
+                if not response:
+                    display.notification('Please enter an {0}.'.format(label), pause=False)
+                else:
+                    return response
+
+    def _get_cloudflare_client(self):
+        """
+        Helper method to construct a `_CloudflareClient`.
+
+        Uses configured credentials.
+
+        :return: a new
+        :rtype: `_CloudflareClient`
+        """
+
+        return _CloudflareClient(self.email, self.api_key)
+
+
+class _CloudflareClient(object):
+    """
+    Encapsulates all communication with the Cloudflare API.
+    """
+
+    def __init__(self, email, api_key):
+        self.cf = CloudFlare.CloudFlare(email, api_key)
+
+    def add_txt_record(self, domain, record_name, record_content, record_ttl):
+        """
+        Add a TXT record using the supplied information.
+
+        :param string domain: The domain to use to look up the Cloudflare zone.
+        :param string record_name: The record name (typically beginning with '_acme-challenge.').
+        :param string record_content: The record content (typically the challenge validation).
+        :param int record_ttl: The record TTL (number of seconds that the record may be cached).
+        :raises: errors.PluginError if an error occurs communicating with the Cloudflare API
+        """
+
+        zone_id = self._find_zone_id(domain)
+
+        data = {'type': 'TXT',
+                'name': record_name,
+                'content': record_content,
+                'ttl': record_ttl}
```

Comment:
> One question I had is about the viability of using the lexicon project you found for many of our DNS plugins including this one. All of the DNS plugins that have been implemented so far are supposedly supported by lexicon with the exception of Google Cloud DNS which they claim to be open to accepting a PR for. If we go this route, I'd also be curious if restructuring our DNS plugins makes sense if there's a common base and level of abstraction from lexicon we can use to easily support all/many of these DNS providers.

In many cases, Lexicon does not provide the functionality we need. Namely: it does not provide functionality for determining the DNS zone which should be modified.

As background: DNS APIs expose the concept of a DNS zone in various ways. Essentially, a DNS zone is a subtree of the overall DNS record tree. The DNS zone for `_acme-challenge.foo.bar.example.com` may be rooted at `_acme-challenge.foo.bar.example.com`, `foo.bar.example.com`, `bar.example.com`, or `example.com` depending on way the provider and the way the user is using the provider. For all DNS APIs I've encountered so far, DNS record creation requests are scoped to a DNS zone. As a result, it is necessary to determine the zone id (called a domain id by some providers).

Most providers have a query-like process for determining the zone id, but that is not exposed by Lexicon.

In some cases, like CloudXNS (see #4522), guess-and-check is the best we can do with or with or without Lexicon. For those, I think using Lexicon makes sense.

For providers that offer a clear query API to determine the zone id, I don't think we should attempt to determine it though guess-and-check (which means not using Lexicon for those providers).

There are also other reasons not to use Lexicon when a better alternative is available:

1. Lexicon often suppresses details when something goes wrong. With `cloudflare`, we can provide distinct error messages for the following cases: invalid credentials, malformed credentials, the domain not being associated with the Cloudflare account, and other errors. With `lexicon`, we would not be able to distinguish those cases.
2. With few exceptions, Lexicon does not use existing API libraries. Using a provider's official library makes it much easier to go to them and ask for help if something breaks. 
3. Lexicon provides a minimum subset of functionality. This is convenient for providing a consistent CLI for managing DNS records. This is less convenient for our use cases; some providers' APIs include functionality to determine whether a record update has been saved or whether a record update has finished propagating (from the provider's point of view). We don't have access to that functionality through Lexicon.

> We split it out into its own package.

If we address #1279, I think this would be worth considering regardless of the dependency discussion; it seems like few users will ever use _all_ DNS plugins. (That said, I think there are arguments to be made in favor of a simple monolithic distribution model that provides a consistent experience for all users.)

> We vendorize the dependencies that aren't widely packaged by OSes such as `cloudflare` and `logger`. I haven't looked into exactly what we'd need to vendorize here but I'm happy to do so if we go this route.

`logging` is just a standard python library, but yes: we could vendorize `cloudflare`.

I don't know much about python dependency management, but after reading a [random article](http://bitprophet.org/blog/2012/06/07/on-vendorizing/), this seems like the right call.